text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,159 @@
+
+If `src` is included in the set passed with that flag, Cargo appends a local registry with path `${$CARGO_RUSTC --print sysroot}/src` to the back of the default chain.
+In other words, if a package is in none of the user-specified registries contain a package, Cargo will look in the registry provided by the compiler.
+
+If `bin` is included in the set passed with that flag (or inferred from the default), Cargo will build a mock registry by examining the contents of the sysroot.
+Any binary in there will be added to the mock registry, with a version deduced the best Cargo can (e.g. from the version of the compiler).
+Cargo likewise will have to be conservative with other metadata, e.g. both aborting if any feature is requested of a dep that is resolved to this mock registry, and also aborting if `default-features = false` is specified in such a dep.
+The mock registry will have dead last priority in the default chain, even behind the source registry.
+The "building" of such a package in the mock registry will consist of copying the binary into the target directory.
+
+[System packages would like to build Rust libraries 1 per system packages, and for this Cargo will need to gain some understanding of prebuilt binaries.
+It is hoped that when it does, the mock registry can be removed and the use of sysroot binaries will be less of a one-off hack.]
+
+Since Cargo will copy any binaries it needs from the sysroot when packages from the mock registry are part of the build plan, it would be nice to always prevent rustc from looking in the sysroot when compiling on Cargo's behalf.
+This would prevent users using the sysroot from forgetting to specify any non-default standard library dependencies.
+A `--no-resolve-sysroot` flag could be added for this purpose, but this is not necessary.
+
+## Rustbuild improvements
+
+When building rust, binaries or source associated with the previous stage are never used, so rustbuild will always pass `--resolve-compiler-specific=` (i.e. that flag with the empty set).
+
+In order to allow building packages that aren't specifically tailored for building rust itself (e.g. they might come from crates.io), Cargo needs to be taught to resolve standard library packages with the current workspace. Either `[[replace]]` can be used for this, or perhaps the members of the workspace would themselves act as a registry.
+
+All binaries for a specific phase can be built with a single `cargo build` (barring special requirements for individual libraries).
+Rather than have a multitude of build artifact directories per stage, only one is needed.
+After the last compiler is build, an additional mini-stage of building just the standard library could be performed, but distributions wishing to build all deps from source in a standardized fashion (e.g. NixOS) would probably forgo this.
```

Comment:
See the thread of diff comments starting with https://github.com/rust-lang/rfcs/pull/1525#discussion_r60848235 for my motivation. I'll try to make this clearer.